src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,462 @@
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed ilk, bytes32 indexed what, address data);
+    event File(bytes32 indexed ilk, bytes32 indexed what, uint256 data);
+    event Wind(bytes32 indexed ilk, uint256 amount);
+    event Unwind(bytes32 indexed ilk, uint256 amount);
+    event Reap(bytes32 indexed ilk, uint256 amt);
+    event Collect(bytes32 indexed ilk, address indexed king, address[] assets, uint256 amt);
+    event Cage();
+    event Cage(bytes32 indexed ilk);
+    event Cull(bytes32 indexed ilk);
+    event Uncull(bytes32 indexed ilk);
+
+    constructor(address chainlog_) public {
+        address vat_ = ChainlogLike(chainlog_).getAddress("MCD_VAT");
+        address daiJoin_ = ChainlogLike(chainlog_).getAddress("MCD_JOIN_DAI");
+
+        chainlog = ChainlogLike(chainlog_);
+        vat = VatLike(vat_);
+        daiJoin = DaiJoinLike(daiJoin_);
+        dai = TokenLike(DaiJoinLike(daiJoin_).dai());
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Math ---
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "DssDirectDepositHub/overflow");
+    }
+    function _sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "DssDirectDepositHub/underflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "DssDirectDepositHub/overflow");
+    }
+    uint256 constant RAY  = 10 ** 27;
+    function _rmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, y) / RAY;
+    }
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, RAY) / y;
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Administration ---
+    function file(bytes32 ilk_, bytes32 what, uint256 data) external auth {
+        Ilk storage ilk = ilks[ilk_];
+        if (what == "tau" ) {
+            require(live == 1, "DssDirectDepositHub/hub-not-live");
+            require(ilk.tic == 0, "DssDirectDepositHub/join-not-live");
+
+            ilk.tau = data;
+        } else revert("DssDirectDepositHub/file-unrecognized-param");
+
+        emit File(ilk_, what, data);
+    }
+
+    function file(bytes32 ilk, bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
+        require(ilks[ilk].tic == 0, "DssDirectDepositHub/pool-not-live");
+
+        if (what == "pool") ilks[ilk].pool = DssDirectDepositPoolLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(ilk, what, data);
+    }
+
+    // --- Deposit controls ---
+    function _wind(bytes32 ilk, DssDirectDepositPoolLike pool, uint256 amount) internal {
```

Comment:
Some of these names are so long... How do people feel about using `D3M` instead of `DssDirectDeposit` when they appear?

Reply:
done https://github.com/makerdao/dss-direct-deposit/pull/31/commits/3eb9e9e07543c58c60b8db17029f1a20764da36c